PLOS/wombat
# Eng 119

https://jira.plos.org/jira/browse/ENG-119

~~Written against the docker branch because that is easier to configure debug logging (to track the number of solr queries) and also because that is the future.~~

Rebased on develop.

Changes:
- Major rewrite of Solr code to ensure that result data is in proper java objects, rather than a lot of untyped code. I also refactored code around to be in more (to me) logical places.
- After this, I was able to identify a few places where solr queries code be reduced:
  1. The faceted queries could be reduced, because multiple facets can be returned from a single query
  1. The desktop search did 2x the searches required, because it uses a dynamic search which threw away the first results returned by the wombat server

Now each search should require only 3 searches, not 10. If memcache is used, 1 of these queries should be loaded from the cache, so there are really only 2 queries made.

The other currently required query is a sub-document faceting query which could probably also be cleaned up.

As far as I can tell, almost all of the time waiting for search results is spent waiting for solr now.

To test:

Run
```
$ mvn package cargo:run
```

Or similar. You should enable the memcached server because the taxonomy browser relies on solr faceting and uses memcached to cache its results, so if you don't have a memcached server you will get poor performance.

~~You should see 3 solr searches for every request to search. ~~

If you want to see how many queries it makes, you'll need to figure out how to enable debug logging or add a `System.out.println` statement to the `SolrSearchApi` class.
